src/main/java/frc/robot/subsystems/drive/Drivetrain.java

```diff
@@ -1,18 +1,427 @@
+  }
+
+  /** stops the drivetrain's movement */
+  public void stop() {
+    move(Vector.Origin, 0);
+  }
+
+  /** runs the periodic functionality of the drivetrain */
   @Override
   public void periodic() {
-    // This method will be called once per scheduler run
+    m_odometry.update(m_gyro.getRotation2d(), m_swerveModulePositions);
+    double ang = getGyroAngle().in(Units.Radians);
+    SmartDashboard.putNumber("delta heading", ang - m_prevAngleRadians);
+
+    m_prevAngleRadians = ang;
+    m_relativeSpeeds = getRobotRelativeSpeeds();
+    m_pose = m_odometry.getPoseMeters();
+
+    SmartDashboard.putNumber("heading", ang - m_headingOffsetRadians);
+
+    SmartDashboard.putNumber("right stick angle", m_rightAngGoalRadians);
+    SmartDashboard.putNumber("turn direction", m_turnDirRadians);
   }
 
-  @Override
-  public void simulationPeriodic() {
-    // This method will be called once per scheduler run during simulation
+  /**
+   * Resets the pose estimator to the specified pose.
+   *
+   * @param pose The pose to which to set the estimator.
+   */
+  public void resetOdometry(Pose2d pose) {
+    m_odometry.resetPosition(
+        Rotation2d.fromRadians(-getGyroAngle().in(Units.Radians)),
+        new SwerveModulePosition[] {
+          m_frontLeft.getPosition(),
+          m_frontRight.getPosition(),
+          m_rearLeft.getPosition(),
+          m_rearRight.getPosition(),
+        },
+        pose);
+  }
+
+  /**
+   * drives the drivatrain using the given inputs the magnitude of the joystick components shouldn't
+   * be > 1 (x^2 + y^2 <= 1)
+   *
+   * @param xSpeed the x-pos of the left joystick (-1, 1)
+   * @param ySpeed the y-pos of the left joystick (-1, 1)
+   * @param xRot the x-pos of the right joystick (-1, 1)
+   * @param yRot the x-pos of the right joystick (-1, 1)
+   * @param altDrive whether or not to use the alternative turning mode
+   * @param centerGyro whether or not to reset the gyro position to the current rotation
+   */
+  public void drive(Vector spdVec, Vector rotVec, boolean altDrive, boolean centerGyro) {
+    if (centerGyro) zeroHeading();
+    if (altDrive) {
+      altDrive(spdVec, rotVec);
+    } else {
+      mainDrive(spdVec, rotVec.x());
+    }
+  }
+
+  /**
+   * moves the divetrain based on the given ChassisSpeeds
+   *
+   * @param spds the target speeds of the drivetrain chassis
+   */
+  public void driveChassisSpeeds(ChassisSpeeds spds) {
+    Vector spd = new Vector(spds.vxMetersPerSecond, spds.vyMetersPerSecond);
+    spdCommanded = spd;
+    double angVel = spds.omegaRadiansPerSecond;
+    move(spd, angVel);
+  }
+
+  /**
+   * moves the drivetrain using the main turning mode
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param xRot the speed to rotate with (-1, 1)
+   */
+  public void mainDrive(Vector spdVec, double xRot) {
```

Comment:
Update your Javadoc

Reply:
Update the Javadoc in what way? Why does it need updated? (When giving feedback, let's practice this as though you are giving it to a new member on the team. You need to give them context, more clear instructions on what it is you are looking for.)